File: app/src/main/java/com/valterc/ki2/karoo/views/Ki2ExtensionView.java
package com.valterc.ki2.karoo.views;

import android.content.Context;
import android.content.res.Configuration;
import android.view.LayoutInflater;
import android.view.View;

import androidx.annotation.NonNull;

import com.valterc.ki2.karoo.Ki2ExtensionContext;
import com.valterc.ki2.karoo.service.ServiceClient;

import io.hammerhead.karooext.models.ViewConfig;

public abstract class Ki2ExtensionView {

    private final Ki2ExtensionContext context;
    private Runnable viewUpdateListener;

    public Ki2ExtensionView(@NonNull Ki2ExtensionContext context) {
        this.context = context;
    }

    @NonNull
    protected Context getContext() {
        return context.getContext();
    }

    @NonNull
    protected Ki2ExtensionContext getKi2ExtensionContext() {
        return context;
    }

    @NonNull
    protected ServiceClient getServiceClient() {
        return context.getServiceClient();
    }

    protected KarooTheme getKarooTheme() {
        int nightModeFlags = getContext().getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK;
        return nightModeFlags == Configuration.UI_MODE_NIGHT_YES ? KarooTheme.DARK : KarooTheme.WHITE;
    }

    protected void viewUpdated() {
        if (viewUpdateListener != null) {
            viewUpdateListener.run();
        }
    }

    @NonNull
    protected abstract View createView(@NonNull LayoutInflater layoutInflater, ViewConfig viewConfig);

    public void setViewUpdateListener(Runnable listener) {
        viewUpdateListener = listener;
    }

    public View createView(ViewConfig viewConfig) {
        return createView(LayoutInflater.from(getContext()), viewConfig);
    }
}

File: app/src/main/java/com/valterc/ki2/karoo/views/Ki2ExtensionView.kt
package com.valterc.ki2.karoo.views

import android.content.Context
import android.content.res.Configuration
import android.view.LayoutInflater
import android.view.View
import com.valterc.ki2.karoo.Ki2ExtensionContext
import com.valterc.ki2.karoo.service.ServiceClient
import io.hammerhead.karooext.models.ViewConfig

abstract class Ki2ExtensionView(protected val extensionContext: Ki2ExtensionContext) {
    private var viewUpdateListener: Runnable? = null

    protected val context: Context
        get() = extensionContext.context

    protected val serviceClient: ServiceClient
        get() = extensionContext.serviceClient

    protected val karooTheme: KarooTheme
        get() {
            val nightModeFlags = context.resources
                .configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
            return if (nightModeFlags == Configuration.UI_MODE_NIGHT_YES) KarooTheme.DARK else KarooTheme.WHITE
        }

    protected fun viewUpdated() {
        viewUpdateListener?.run()
    }

    protected abstract fun createView(layoutInflater: LayoutInflater, viewConfig: ViewConfig): View

    open fun dispose() {
        setViewUpdateListener(null)
    }

    fun setViewUpdateListener(listener: Runnable?) {
        viewUpdateListener = listener
    }

    fun createView(viewConfig: ViewConfig): View {
        return createView(LayoutInflater.from(context), viewConfig)
    }
}
